File: DungeonDebt/Assets/Scripts/Run/ShopManager.cs
using System.Collections.Generic;
using UnityEngine;

public class ShopManager : MonoBehaviour
{
    private const int NoEmptyFormationSlot = -1;

    [SerializeField] private RunManager _runManager;

    private readonly List<ShopOffer> _currentOffers = new List<ShopOffer>();

    public IReadOnlyList<ShopOffer> CurrentOffers
    {
        get { return _currentOffers; }
    }

    public void Initialize(RunManager runManager)
    {
        _runManager = runManager;
    }

    public void GenerateOffers()
    {
        FillAllOffers();
    }

    public bool Reroll()
    {
        RunState run = GetRunState();
        if (run == null || run.Gold < GameRules.RerollCost)
        {
            return false;
        }

        run.Gold -= GameRules.RerollCost;
        run.RerollCount += 1;
        FillAllOffers();
        return true;
    }

    public bool Hire(int offerIndex)
    {
        if (offerIndex < 0 || offerIndex >= _currentOffers.Count)
        {
            return false;
        }

        ShopOffer offer = _currentOffers[offerIndex];
        if (offer == null || offer.Purchased)
        {
            return false;
        }

        RunState run = GetRunState();
        if (run == null)
        {
            return false;
        }

        if (run.Gold < offer.HireCost)
        {
            return false;
        }

        HeroInstance existing = FindExistingPartyMember(run, offer.Hero);
        if (existing != null)
        {
            // Defensive: Silver-owned heroes are excluded from the offer pool, so this branch
            // should be unreachable. Guard anyway per the M9.1 plan.
            if (existing.Tier == HeroTier.Silver)
            {
                return false;
            }

            run.Gold -= offer.HireCost;
            existing.Tier = HeroTier.Silver;
            HeroEffects.ApplyTierStatSeed(existing);
            existing.CurrentHealth = HeroEffects.GetTierAdjustedMaxHealth(existing);
            offer.Purchased = true;
            return true;
        }

        if (run.Party.Count >= GameRules.MaxPartySize)
        {
            return false;
        }

        int formationSlot = FindFirstEmptyFormationSlot(run);
        if (formationSlot == NoEmptyFormationSlot)
        {
            return false;
        }

        run.Gold -= offer.HireCost;
        HeroInstance hired = new HeroInstance(offer.Hero, formationSlot);
        hired.Tier = offer.Tier;
        HeroEffects.ApplyTierStatSeed(hired);
        hired.CurrentHealth = HeroEffects.GetTierAdjustedMaxHealth(hired);
        run.Party.Add(hired);
        offer.Purchased = true;
        return true;
    }

    private static HeroInstance FindExistingPartyMember(RunState run, HeroDefinition hero)
    {
        if (run == null || hero == null)
        {
            return null;
        }

        for (int i = 0; i < run.Party.Count; i++)
        {
            HeroInstance member = run.Party[i];
            if (member != null && member.Definition != null && member.Definition.Id == hero.Id)
            {
                return member;
            }
        }

        return null;
    }

    public bool Fire(int partyIndex)
    {
        RunState run = GetRunState();
        if (run == null)
        {
            return false;
        }

        if (partyIndex < 0 || partyIndex >= run.Party.Count)
        {
            return false;
        }

        run.Party.RemoveAt(partyIndex);
        run.Gold += GameRules.FireRefund;

        for (int i = 0; i < run.Party.Count; i++)
        {
            run.Party[i].FormationSlot = i;
        }

        return true;
    }

    private void FillAllOffers()
    {
        _currentOffers.Clear();

        System.Random rng = _runManager != null ? _runManager.Random : null;
        if (rng == null)
        {
            return;
        }

        RunState run = GetRunState();
        HashSet<string> silverOwned = new HashSet<string>();
        HashSet<string> anyOwned = new HashSet<string>();
        if (run != null)
        {
            for (int i = 0; i < run.Party.Count; i++)
            {
                HeroInstance member = run.Party[i];
                if (member == null || member.Definition == null)
                {
                    continue;
                }

                anyOwned.Add(member.Definition.Id);
                if (member.Tier == HeroTier.Silver)
                {
                    silverOwned.Add(member.Definition.Id);
                }
            }
        }

        IReadOnlyList<HeroDefinition> allHeroes = DataRepository.AllHeroes;

        // Bronze pool: heroes whose Silver upgrade is not already owned. Bronze-owned
        // heroes stay in the pool so a duplicate hire can upgrade them to Silver.
        List<HeroDefinition> bronzePool = new List<HeroDefinition>();
        // Silver pool: only heroes the player does not own at all. A Silver offer for
        // a Bronze-owned hero would be paying extra for the same merge outcome, so
        // exclude them and let the Bronze duplicate path handle the upgrade.
        List<HeroDefinition> silverPool = new List<HeroDefinition>();
        for (int i = 0; i < allHeroes.Count; i++)
        {
            HeroDefinition hero = allHeroes[i];
            if (!silverOwned.Contains(hero.Id))
            {
                bronzePool.Add(hero);
            }
            if (!anyOwned.Contains(hero.Id))
            {
                silverPool.Add(hero);
            }
        }

        for (int i = 0; i < GameRules.ShopOfferCount; i++)
        {
            bool wantSilver = rng.NextDouble() < GameRules.SilverOfferChance;
            List<HeroDefinition> pool = wantSilver && silverPool.Count > 0 ? silverPool : bronzePool;
            HeroTier tier = (pool == silverPool) ? HeroTier.Silver : HeroTier.Bronze;

            if (pool.Count == 0)
            {
                _currentOffers.Add(null);
                continue;
            }

            int pick = rng.Next(0, pool.Count);
            HeroDefinition picked = pool[pick];
            pool.RemoveAt(pick);
            // Keep the other pool consistent so the same hero can't double-appear.
            if (tier == HeroTier.Silver)
            {
                RemoveById(bronzePool, picked.Id);
            }
            else
            {
                RemoveById(silverPool, picked.Id);
            }

            int hireCost = picked.BaseUpkeep + GameRules.HireCostBonus;
            if (tier == HeroTier.Silver || anyOwned.Contains(picked.Id))
            {
                hireCost += GameRules.SilverHireCostBonus;
            }

            _currentOffers.Add(new ShopOffer(picked, hireCost, tier));
        }
    }

    private static void RemoveById(List<HeroDefinition> pool, string heroId)
    {
        for (int i = 0; i < pool.Count; i++)
        {
            if (pool[i].Id == heroId)
            {
                pool.RemoveAt(i);
                return;
            }
        }
    }

    private RunState GetRunState()
    {
        if (_runManager == null)
        {
            return null;
        }

        return _runManager.CurrentRunState;
    }

    private static int FindFirstEmptyFormationSlot(RunState run)
    {
        if (run == null)
        {
            return NoEmptyFormationSlot;
        }

        for (int slot = 0; slot < GameRules.MaxPartySize; slot++)
        {
            bool occupied = false;
            for (int i = 0; i < run.Party.Count; i++)
            {
                if (run.Party[i].FormationSlot == slot)
                {
                    occupied = true;
                    break;
                }
            }

            if (!occupied)
            {
                return slot;
            }
        }

        return NoEmptyFormationSlot;
    }
}

File: web/src/run/ShopManager.js
// Ported from DungeonDebt/Assets/Scripts/Run/ShopManager.cs
import { HeroTier, ShopEventId } from "../data/enums.js";
import { GameRules, GameRulesFns } from "../core/GameRules.js";
import { DataRepository } from "../core/DataRepository.js";
import { HeroInstance } from "../data/HeroInstance.js";
import { ShopOffer } from "../data/ShopOffer.js";
import { HeroEffects } from "../combat/HeroEffects.js";

const NoEmptyFormationSlot = -1;

export class ShopManager {
  constructor(runManager = null) {
    this._runManager = runManager;
    this._currentOffers = [];
  }

  get currentOffers() { return this._currentOffers; }

  initialize(runManager) {
    this._runManager = runManager;
  }

  generateOffers() {
    this._fillAllOffers();
  }

  clearShopEvent() {
    const run = this._getRunState();
    if (run) run.currentShopEvent = null;
  }

  reroll() {
    const run = this._getRunState();
    if (!run || run.gold < GameRules.RerollCost) return false;
    run.gold -= GameRules.RerollCost;
    run.rerollCount += 1;
    this._fillAllOffers();
    return true;
  }

  payDebt() {
    const run = this._getRunState();
    if (!run) return false;
    const payment = GameRulesFns.calculateDebtPaymentAmount(run.gold, run.debt);
    if (payment <= 0) return false;
    run.gold -= payment;
    run.debt -= payment;
    return true;
  }

  hire(offerIndex) {
    if (offerIndex < 0 || offerIndex >= this._currentOffers.length) return false;
    const offer = this._currentOffers[offerIndex];
    if (!offer || offer.purchased) return false;

    const run = this._getRunState();
    if (!run) return false;
    if (run.gold < offer.hireCost) return false;

    const existing = findExistingPartyMember(run, offer.hero);
    if (existing) {
      if (existing.tier === HeroTier.Diamond) return false;
      run.gold -= offer.hireCost;
      existing.tier = getNextMergeTier(existing.tier);
      HeroEffects.applyTierStatSeed(existing);
      existing.currentHealth = HeroEffects.getTierAdjustedMaxHealth(existing);
      offer.purchased = true;
      return true;
    }

    if (run.party.length >= GameRules.MaxPartySize) return false;

    const formationSlot = findFirstEmptyFormationSlot(run);
    if (formationSlot === NoEmptyFormationSlot) return false;

    run.gold -= offer.hireCost;
    const hired = new HeroInstance(offer.hero, formationSlot);
    hired.tier = offer.tier;
    HeroEffects.applyTierStatSeed(hired);
    hired.currentHealth = HeroEffects.getTierAdjustedMaxHealth(hired);
    run.party.push(hired);
    offer.purchased = true;
    return true;
  }

  fire(partyIndex) {
    const run = this._getRunState();
    if (!run) return false;
    if (partyIndex < 0 || partyIndex >= run.party.length) return false;

    run.party.splice(partyIndex, 1);
    run.gold += GameRules.FireRefund;

    for (let i = 0; i < run.party.length; i++) {
      run.party[i].formationSlot = i;
    }
    return true;
  }

  _fillAllOffers() {
    this._currentOffers.length = 0;

    const rng = this._runManager ? this._runManager.rng : null;
    if (!rng) return;

    const run = this._getRunState();
    const terminalOwned = new Set();
    const anyOwned = new Set();
    if (run) {
      for (const member of run.party) {
        if (!member || !member.definition) continue;
        anyOwned.add(member.definition.id);
        if (member.tier === HeroTier.Diamond) terminalOwned.add(member.definition.id);
      }
    }

    const allHeroes = DataRepository.allHeroes;

    // Bronze pool: heroes not already at Diamond (can still be merged up).
    const bronzePool = [];
    // Silver pool: only heroes the player does not own at all.
    const silverPool = [];
    for (const hero of allHeroes) {
      if (!terminalOwned.has(hero.id)) bronzePool.push(hero);
      if (!anyOwned.has(hero.id)) silverPool.push(hero);
    }

    for (let i = 0; i < GameRules.ShopOfferCount; i++) {
      const wantSilver = rng.nextDouble() < GameRules.SilverOfferChance;
      const pool = wantSilver && silverPool.length > 0 ? silverPool : bronzePool;
      const tier = pool === silverPool ? HeroTier.Silver : HeroTier.Bronze;

      if (pool.length === 0) {
        this._currentOffers.push(null);
        continue;
      }

      const pick = rng.next(0, pool.length);
      const picked = pool[pick];
      pool.splice(pick, 1);
      // Keep the other pool consistent so the same hero can't double-appear.
      if (tier === HeroTier.Silver) removeById(bronzePool, picked.id);
      else removeById(silverPool, picked.id);

      let hireCost = picked.baseUpkeep + GameRules.HireCostBonus;
      if (tier === HeroTier.Silver || anyOwned.has(picked.id)) {
        hireCost += GameRules.SilverHireCostBonus;
      }

      this._currentOffers.push(new ShopOffer(picked, hireCost, tier));
    }

    // M17 — Roll a shop event (~20%).
    this._rollShopEvent(run);
  }

  _rollShopEvent(run) {
    if (!run) return;
    run.currentShopEvent = null;

    const rng = this._runManager ? this._runManager.rng : null;
    if (!rng) return;

    // 20% chance for any event
    if (rng.nextDouble() >= 0.2) return;

    // Pick one event from the pool
    const eventPool = [ShopEventId.BargainStall, ShopEventId.TaxAudit, ShopEventId.TravellingMerchant];
    const eventId = eventPool[rng.next(eventPool.length)];

    if (eventId === ShopEventId.BargainStall) {
      this._applyBargainStallEvent(run);
    } else if (eventId === ShopEventId.TaxAudit) {
      run.currentShopEvent = { eventId: ShopEventId.TaxAudit };
    } else if (eventId === ShopEventId.TravellingMerchant) {
      run.currentShopEvent = {
        eventId: ShopEventId.TravellingMerchant,
        purchases: [],
        goods: [
          { id: "healAll", label: "Heal All Party", cost: GameRules.TravellingHealAllCost, description: "Restore all heroes to full HP" },
          { id: "goldBlessing", label: "Gold Blessing", cost: GameRules.TravellingBlessingCost, description: "+" + GameRules.TravellingBlessingAmount + " gold on next combat reward" },
        ],
      };
    }
  }

  _applyBargainStallEvent(run) {
    const validSlots = [];
    for (let i = 0; i < this._currentOffers.length; i++) {
      if (this._currentOffers[i]) validSlots.push(i);
    }
    if (validSlots.length === 0) return;

    const rng = this._runManager ? this._runManager.rng : null;
    if (!rng) return;

    const slotIndex = validSlots[rng.next(validSlots.length)];
    const offer = this._currentOffers[slotIndex];
    const originalCost = offer.hireCost;
    const discountedCost = Math.max(1, Math.ceil(originalCost * 0.5));

    offer.hireCost = discountedCost;

    run.currentShopEvent = {
      eventId: ShopEventId.BargainStall,
      slotIndex,
      originalCost,
      discountedCost,
    };
  }

  resolveTaxAudit(payGold) {
    const run = this._getRunState();
    if (!run || !run.currentShopEvent || run.currentShopEvent.eventId !== ShopEventId.TaxAudit) return false;

    if (payGold) {
      if (run.gold < GameRules.TaxAuditGoldCost) return false;
      run.gold -= GameRules.TaxAuditGoldCost;
    } else {
      run.morale -= 1;
    }
    run.currentShopEvent = null;
    return true;
  }

  purchaseTravellingGood(itemId) {
    const run = this._getRunState();
    if (!run || !run.currentShopEvent || run.currentShopEvent.eventId !== ShopEventId.TravellingMerchant) return false;

    const goods = run.currentShopEvent.goods;
    const good = goods.find((g) => g.id === itemId);
    if (!good) return false;
    if (this.isTravellingGoodPurchased(itemId)) return false;
    if (run.gold < good.cost) return false;

    run.gold -= good.cost;
    run.currentShopEvent.purchases.push(itemId);

    if (itemId === "healAll") {
      for (const hero of run.party) {
        hero.currentHealth = HeroEffects.getTierAdjustedMaxHealth(hero);
      }
    } else if (itemId === "goldBlessing") {
      run.pendingNextRewardBonus += GameRules.TravellingBlessingAmount;
    }

    return true;
  }

  isTravellingGoodPurchased(itemId) {
    const run = this._getRunState();
    if (!run || !run.currentShopEvent || run.currentShopEvent.eventId !== ShopEventId.TravellingMerchant) return false;
    return run.currentShopEvent.purchases.indexOf(itemId) !== -1;
  }

  _getRunState() {
    if (!this._runManager) return null;
    return this._runManager.currentRunState;
  }
}

function getNextMergeTier(tier) {
  if (tier === HeroTier.Bronze) return HeroTier.Silver;
  if (tier === HeroTier.Silver) return HeroTier.Gold;
  if (tier === HeroTier.Gold) return HeroTier.Diamond;
  return tier;
}

function findExistingPartyMember(run, hero) {
  if (!run || !hero) return null;
  for (const member of run.party) {
    if (member && member.definition && member.definition.id === hero.id) return member;
  }
  return null;
}

function removeById(pool, heroId) {
  for (let i = 0; i < pool.length; i++) {
    if (pool[i].id === heroId) {
      pool.splice(i, 1);
      return;
    }
  }
}

function findFirstEmptyFormationSlot(run) {
  if (!run) return NoEmptyFormationSlot;
  for (let slot = 0; slot < GameRules.MaxPartySize; slot++) {
    let occupied = false;
    for (const member of run.party) {
      if (member.formationSlot === slot) {
        occupied = true;
        break;
      }
    }
    if (!occupied) return slot;
  }
  return NoEmptyFormationSlot;
}
